cps/main.c

```diff
@@ -16,11 +16,521 @@
+			rte_pktmbuf_free(bufs[i]);
+	}
+
+	/*
+	 * Userspace requests to change the device MTU or configure the
+	 * device up/down are forwarded from the kernel back to userspace
+	 * for DPDK to handle. rte_kni_handle_request() receives those
+	 * requests and allows them to be processed.
+	 */
+	if (rte_kni_handle_request(kni) < 0)
+		RTE_LOG(WARNING, KNI,
+			"%s: error in handling userspace request on KNI %s\n",
+			__func__, rte_kni_get_name(kni));
+}
+
+static void
+process_egress(struct gatekeeper_if *iface, struct rte_kni *kni,
+	uint16_t tx_queue)
+{
+	struct rte_mbuf *bufs[GATEKEEPER_MAX_PKT_BURST];
+	uint16_t num_rx = rte_kni_rx_burst(kni, bufs, GATEKEEPER_MAX_PKT_BURST);
+	unsigned int num_tx = rte_eth_tx_burst(iface->id, tx_queue,
+		bufs, num_rx);
+
+	/*
+	 * TODO Forward BGP, respond to ARP or ND requests with resolution,
+	 * and drop everything else.
+	 */
+
+	if (unlikely(num_tx < num_rx)) {
+		uint16_t i;
+		for (i = num_tx; i < num_rx; i++)
+			rte_pktmbuf_free(bufs[i]);
+	}
+}
+
+static int
+cps_proc(void *arg)
+{
+	struct cps_config *cps_conf = (struct cps_config *)arg;
+	struct net_config *net_conf = cps_conf->net;
+
+	struct gatekeeper_if *front_iface = &net_conf->front;
+	struct gatekeeper_if *back_iface = &net_conf->back;
+	struct rte_kni *front_kni = cps_conf->front_kni;
+	struct rte_kni *back_kni = cps_conf->back_kni;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"cps: the CPS block is running at lcore = %u\n",
+		cps_conf->lcore_id);
+
+	while (likely(!exiting)) {
+		/*
+		 * Read in IPv4 BGP packets that arrive directly
+		 * on the Gatekeeper interfaces.
+		 */
+		process_ingress(front_iface, front_kni,
+			cps_conf->rx_queue_front);
+		if (net_conf->back_iface_enabled)
+			process_ingress(back_iface, back_kni,
+				cps_conf->rx_queue_back);
+
+		/*
+		 * Process any requests made to the CPS block, including
+		 * IPv6 BGP packets that arrived via an ACL.
+		 */
+		process_reqs(cps_conf);
+
+		/*
+		 * Read in packets from KNI interfaces, and
+		 * transmit to respective Gatekeeper interfaces.
+		 */
+		process_egress(front_iface, front_kni,
+			cps_conf->tx_queue_front);
+		if (net_conf->back_iface_enabled)
+			process_egress(back_iface, back_kni,
+				cps_conf->tx_queue_back);
+
+		/* TODO Get route updates from kernel. */
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"cps: the CPS block at lcore = %u is exiting\n",
+		cps_conf->lcore_id);
+
+	return cleanup_cps();
+}
+
+static int
+submit_bgp(struct rte_mbuf **pkts, int num_pkts, struct gatekeeper_if *iface)
+{
+	struct cps_config *cps_conf = get_cps_conf();
+	struct cps_request *req = mb_alloc_entry(&cps_conf->mailbox);
+	int ret;
+
+	if (req == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: allocation of mailbox message failed\n");
+		return -1;
+	}
+
+	req->ty = CPS_REQ_BGP;
+	req->u.bgp.num_pkts = num_pkts;
+	req->u.bgp.kni = iface == &cps_conf->net->front
+		? cps_conf->front_kni
+		: cps_conf->back_kni;
+	rte_memcpy(req->u.bgp.pkts, pkts, sizeof(*req->u.bgp.pkts) * num_pkts);
```

Comment:
For (2), adding an `RTE_VERIFY()` on `num_pkts` caused a compilation error for comparing signed and unsigned numbers, so I went back to the **acl: improve IPv6 ACL config** patch that you marked as ready and changed the ACL callback interface so that the number of packets is `unsigned int`, not `int`.

Therefore, that patch will show up as updated in the rebase, but that's the only change that I made to it.